comit/src/expiries.rs

```diff
@@ -0,0 +1,631 @@
+        let now = self.beta_connector.current_time();
+
+        // Alice redeems on beta ledger so is concerned about the beta expiry.
+        let end_time = now.add_duration(period);
+        end_time < self.beta_expiry
+    }
+
+    /// True if Bob has time to complete a swap (i.e transition to done).
+    fn bob_has_time_to_complete(&self, current_state: BobState) -> bool {
+        let period = period_for_bob_to_complete(&self.config, current_state);
+        let now = self.alpha_connector.current_time();
+
+        // Bob redeems on alpha ledger so is concerned about the alpha expiry.
+        let end_time = now.add_duration(period);
+        end_time < self.alpha_expiry
+    }
+
+    /// If Alice's next action is not taken within X minutes the expiries will
+    /// become un-useful. Returns X.
+    pub fn alice_should_act_within(&self, current_state: AliceState) -> Duration {
+        let period = period_for_alice_to_complete(&self.config, current_state);
+        let start_time = self.beta_expiry.sub_duration(period);
+        let now = self.beta_connector.current_time();
+
+        timestamp::duration_between(now, start_time)
+    }
+
+    /// If Bob's next action is not taken within X minutes the expiries will
+    /// become un-useful. Returns X.
+    pub fn bob_should_act_within(&self, current_state: BobState) -> Duration {
+        let period = period_for_bob_to_complete(&self.config, current_state);
+        let start_time = self.alpha_expiry.sub_duration(period);
+        let now = self.alpha_connector.current_time();
+
+        timestamp::duration_between(now, start_time)
+    }
+
+    /// Returns the recommended next action that Alice should take.
+    pub fn next_action_for_alice(&self, current_state: AliceState) -> AliceAction {
+        if current_state == AliceState::Done {
+            return AliceAction::NoFurtherAction;
+        }
+
+        let funded = current_state.has_funded_alpha();
+
+        if self.alpha_expiry_has_elapsed() {
+            if funded {
+                return AliceAction::Refund;
+            }
+            return AliceAction::Abort;
+        }
+
+        if !self.alice_has_time_to_complete(current_state) {
+            if funded {
+                return AliceAction::WaitForRefund;
+            }
+            return AliceAction::Abort;
+        }
+
+        let (next_action, _state) = current_state.next();
+        next_action
+    }
+
+    /// Returns the recommended next action that Bob should take.
+    pub fn next_action_for_bob(&self, current_state: BobState) -> BobAction {
+        if current_state == BobState::Done {
+            return BobAction::NoFurtherAction;
+        }
+
+        let funded = current_state.has_funded_beta();
+
+        if self.beta_expiry_has_elapsed() {
+            if funded {
+                return BobAction::Refund;
+            }
+            return BobAction::Abort;
+        };
+
+        if !self.bob_has_time_to_complete(current_state) {
```

Comment:
📣 This does not take in account if Alice has redeemed. If she has redeemed, Bob should redeem, not matter the current time.

Reply:
Nice catch! Thanks man